staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/cbor.go

```diff
@@ -0,0 +1,111 @@
+package v1
+
+import (
+	"errors"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+func IsMapOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	dst := new(struct{})
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) && uterr.GoType == "string" {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func IsArrayOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	var dst [0]any
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) && uterr.GoType == "string" {
+			return true
+		}
+		return false
+	}
+
+	return true
+}
+
+func IsStringArrayOrNull(in []byte) bool {
+	decMode, err := cbor.DecOptions{
+		DupMapKey:           cbor.DupMapKeyQuiet,
+		FieldNameMatching:   cbor.FieldNameMatchingCaseSensitive,
+		FieldNameByteString: cbor.FieldNameByteStringAllowed,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	var dst [0]string
+	if err := decMode.Unmarshal(in, &dst); err != nil {
+		var uterr *cbor.UnmarshalTypeError
+		if errors.As(err, &uterr) {
+			if uterr.GoType == "string" {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+type cborDataItemType int
+
+const (
+	cborNullType cborDataItemType = iota
+	cborBoolType
+	cborNumberType
+	cborStringType
+	cborArrayType
+	cborMapType
+	cborTagType
+)
+
+func cborType(b byte) cborDataItemType {
```

Comment:
Following from my comment above: as a caller, I'd expect `cborType` to return one of the eight major types. As a side effect of making the contract stricter, I think the implementation becomes easier to understand, too, like:

```go
const (
  // https://www.rfc-editor.org/rfc/rfc8949.html#section-3.1
  cborUnsignedInteger cborMajorType = 0
  cborNegativeInteger = 1
  cborByteString = 2
  cborTextString = 3
  cborArray = 4
  cborMap = 5
  cborTag = 6
  cborOther = 7
)

func cborType(initialByte byte) cborMajorType {
  return initialByte >> 5 // major type is encoded in the 3 most significant bits of the initial byte of a data item
}
```

Reply:
Sorry, the linter is complaining because I should not have dropped the types from these consts in my suggested change.